---
title: "Class09"
author: "Yuntian Zhu (PID: A17816597)"
format: html
---

The main database for structural biology is PDB. Let us take a look at what it contains:

Download a CSV file from the PDB site (accessible from “Analyze” > “PDB Statistics” > “by Experimental Method and Molecular Type”. Move this CSV file into your RStudio project and use it to answer the following questions:

First, add the csc file to the R code.

```{r}
data <- read.csv("Data Export Summary.csv")
data
```

```{r}
data$Total
```

Note that the numbers here are indeed strings.

We can use a package to fix this problem. 

```{r pdbstats, message = FALSE}
library(readr)
stats <- read_csv("Data Export Summary.csv")
stats
```


>Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
n.total <- sum(stats$Total)
n.xray <- sum(stats$`X-ray`)
n.EM <- sum(stats$EM)

#Use the round function to round to 2 significant figures
round((n.xray/n.total)*100, 2)
round((n.EM/n.total)*100, 2)
```

Therefore, 81.48% of the total structures are solved by X-ray; 12.22% of the total structures are solved by EM.

>Q2: What proportion of structures in the PDB are protein?

```{r}
protein_only <- stats[stats$`Molecular Type` == "Protein (only)", ]
p.total <- protein_only$Total
round((p.total/n.total)*100, 2)

protein <- stats[stats$`Molecular Type` == "Protein (only)" | 
                   stats$`Molecular Type`=="Protein/Oligosaccharide" |
                   stats$`Molecular Type`=="Protein/NA", ]
round((sum(protein$Total)/n.total)*100, 2)
```

Therefore, 86.05% of the total structures are protein only; 97.62% are protein only or protein combined with nucleic acids/oligosaccharide.

>Q3: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

A seearch with the key words "HIV-I" and "Protease" results in 1002 structures (see the picture below, the number is in the red box)

![](Search.png)

**Package for structural bioinformatics**

```{r}
library(bio3d)
hiv <- read.pdb("1hsg")
hiv
```
Let us first use the Mol* viewer to explore this structure

![My first view of HIV-Pr](1HSG.png)
And a view of the ligand with catalytic ASP 25 amino acids and the all important active site water molecule. The ASP 25 and water molecules are represented with the spacefill model. 

![](1HSG-2.png)

**PDB objects in R**

```{r}
library(bio3d)
hiv <- read.pdb("1hsg")
head(hiv$atom)
```

Extract the sequence

```{r}
pdbseq(hiv)
```

```{r}
chainA_seq <- pdbseq(trim.pdb(hiv, chain = "A"))
chainA_seq
```

I can interactively view these PDB objects in R with the new **bio3dview**. This package is not in CRAN yet since it is new. 

```{r}
library(bio3dview)
library(NGLVieweR)
view.pdb(hiv)
```

Change some settings

```{r}
sel <- atom.select(hiv, resno = 25)
view.pdb(hiv, highlight = sel,
         highlight.style = "spacefill",
         colorScheme = "chain", 
         col = c("blue", "orange"),
         backgroundColor = "pink")
```

**Predict protein flexibility**

We can run a bioinformatis calculation to predict protein dynamics -- i. e. functional motions.

We will use the `nma()` function:

```{r}
adk <- read.pdb("6s36")
adk
```

```{r}
m <- nma(adk)
m
plot(m)
```
Generate a "trajectory" of predicted motion

```{r}
mktrj(m, file = "ADK_nma.pdb")
```

```{r}
view.nma(m)
```

**Comparative structure analysis of Adenylate Kinase**

> Q10. Which of the packages above is found only on BioConductor and not CRAN? 

msa

> Q11. Which of the above packages is not found on BioConductor or CRAN?: 

bioboot/bio3dview

> Q12. True or False? Functions from the pak package can be used to install packages from GitHub and BitBucket? 

True

Now, let us first get the the query sequence for chain A of the PDB ID 1AKE

```{r}
aa <- get.seq("1ake_A")
aa
```

> Q13. How many amino acids are in this sequence, i.e. how long is this sequence?

214.

Then, we can perform a blast search

```{r}
# Blast or hmmer search 
# b <- blast.pdb(aa)
# Plot a summary of search results
# hits <- plot(b)
# List out some 'top hits'
# head(hits$pdb.id)
# The TRUE BLAST takes too long time, especially when rendering the file. 
hits <- NULL
hits$pdb.id <- c('1AKE_A','6S36_A','6RZE_A',
                 '3HPR_A','1E4V_A','5EJE_A',
                 '1E4Y_A','3X2S_A','6HAP_A',
                 '6HAM_A','4K46_A','3GMT_A','4PZL_A')
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

Next we will use the pdbaln() function to align and also optionally fit (i.e. superpose) the identified PDB structures.

```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)
# Draw schematic alignment
plot(pdbs, labels=ids)
```

In addition, we can annotate collected PDB structures

```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
anno
```

Finally, we can analyze the data by performing the PCA analysis

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

